securedrop_client/gui/widgets.py

```diff
@@ -961,20 +958,41 @@ def update(self, sources: List[Source]) -> List[str]:
         return deleted_uuids
 
     def get_current_source(self):
-        source_item = self.currentItem()
-        source_widget = self.itemWidget(source_item)
-        if source_widget and source_exists(self.controller.session, source_widget.source_uuid):
-            return source_widget.source
+        try:
+            source_item = self.currentItem()
+            source_widget = self.itemWidget(source_item)
+            if source_widget:
+                return self.controller.get_source(source_widget.source_uuid)
+        except sqlalchemy.exc.InvalidRequestError:
+            pass
 
-    def set_snippet(self, source_uuid, message_uuid, content):
-        """
-        Given a UUID of a source, if the referenced message is the latest
-        message, then update the source's preview snippet to the referenced
-        content.
-        """
-        source_widget = self.source_widgets.get(source_uuid)
+    def get_source_widget(self, source_uuid: str) -> QListWidget:
+        '''
+        First try to get the source widget from the cache, then look for it in the SourceList.
+        '''
+        try:
+            source_widget = self.source_widgets[source_uuid]
+            return source_widget
+        except KeyError:
+            pass
+
+        for i in range(self.count()):
+            list_item = self.item(i)
+            source_widget = self.itemWidget(list_item)
+            if source_widget and source_widget.source_uuid == source_uuid:
+                return source_widget
```

Comment:
clarifying: under which circumstances can the source widget be in the cache but not the source list?

Reply:
I think we could rely completely on the cache. I abstracted getting the source widget here because I was thinking we didn't need both the qt list and the cache since we have to loop through all items in the qt list (`SourceList`) anyway. But after thinking about this more, we have future plans to modify server code so that it only returns sources that have updates, and that's when the cache will come in handy (because we won't have to loop through the entire source list anymore and update all the source widgets). Long story short: I can update this method so that there's no fall back if we try to get a source widget that hasn't been added to the cache yet (since we add items to the cache before adding them to the qt list, this currently can't happen but if we ever switch the order of this by accident, this method will ensure we'll be able to get the selected source widget)